### PR TITLE
Add option to run External through a CUDA graph 

### PR DESCRIPTION
**Pull request by RaulPPelaez**
This PR allows to capture a TorchMD-Net model to run it in TorchMD.

A new argument is added to external, use_cuda_graphs that defaults to False

### CODE REVIEW

**Comment by RaulPPelaez** [comment]
@AntonioMirarchi this is done and GPU tests pass for me. Could you give it a try?

**Comment by RaulPPelaez** [comment]
Using TensorNet in TorchMD with CUDA-graphs enabled gave me an error in an apparently innocuous operation:
```python
  File "/shared/raul/torchmd-net/torchmdnet/models/tensornet.py", line 59, in new_radial_tensor                                                                                                    
    In = f_I[..., None, None] * I                                                                                                                                                                  
RuntimeError: CUDA error: operation not permitted when stream is capturing                     
```
I had to explicitly use intermediate tensors like this for TorchMD to CUDA-graph succesfully.
```diff
 # Modifies tensor by multiplying invariant features to irreducible components
 def new_radial_tensor(I, A, S, f_I, f_A, f_S):
-    I = f_I[..., None, None] * I
-    A = f_A[..., None, None] * A
-    S = f_S[..., None, None] * S
+    # Writing this function without using intermediate tensors breaks CUDA graphs in some cases
+    tempf = f_I[..., None, None]
+    I =  tempf * I
+    tempf = f_A[..., None, None]
+    A = tempf * A
+    tempf = f_S[..., None, None]
+    S = tempf * S
```  
This makes no sense to me and I find it quite scary, torch is a fickle mistress...
cc @raimis in case you encounter this or have some insight.

**Comment by RaulPPelaez** [comment]
@AntonioMirarchi this is working for you, right?

**Comment by AntonioMirarchi** [comment]
Yes, I'm using it

**Comment by RaulPPelaez** [comment]
@raimis @AntonioMirarchi please review

File: torchmdnet/calculators.py [hunk 1]
```diff
@@ -46,11 +67,46 @@ def __init__(self, netfile, embeddings, device="cpu", output_transform=None):
...
+                self._init_cuda_graph()
+            with torch.no_grad():
+                self.pos.copy_(pos)
+                self.cuda_graph.replay()
+        else:
+            self.energy, self.forces = self.model(self.embeddings, pos, self.batch)
         return self.output_transformer(
-            energy.detach(), forces.reshape(-1, self.n_atoms, 3).detach()
+            self.energy.clone().detach(),
+            self.forces.clone().reshape(-1, self.n_atoms, 3).detach(),
```
[the remark this answers is not in the document]
Comment (RaulPPelaez) [reply torchmdnet/calculators.py:111]: Do we need External to always return the force no matter what?
It is actually more performant to say derivative=False and do the backwards in External. This way one avoids running the backwards pass with autograd enabled, as it is the case in TorchMD_Net.

[the remark this answers is not in the document]
Comment (RaulPPelaez) [reply torchmdnet/calculators.py:111]: Alternatively, we can put the model in eval mode and detect it in TorchMD_Net:

```python
        if self.derivative:
            grad_outputs: List[Optional[torch.Tensor]] = [torch.ones_like(y)]
            dy = grad(
                [y],
                [pos],
                grad_outputs=grad_outputs,
                create_graph=self.training,
                retain_graph=self.training,
            )[0]
            if dy is None:
                raise RuntimeError("Autograd returned None for the force prediction.")
```
This is probably the cleanest solution here.

**Review by AntonioMirarchi — APPROVED** [review]

**Review by stefdoerr — APPROVED** [review]

**Pull request merged**